plugins/inputs/windows_event_log/wineventlog/wineventlog_test.go

```diff
@@ -0,0 +1,184 @@
+	assert.Zero(t, elog.eventHandle)
+}
+
+// TestOpen verifies Open() succeeds with valid inputs.
+// And fails with invalid inputs.
+func TestOpen(t *testing.T) {
+	// Happy path.
+	elog := NewEventLog(NAME, LEVELS, GROUP_NAME, STREAM_NAME, RENDER_FMT, DEST,
+		STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.NoError(t, elog.Open())
+	assert.NotZero(t, elog.eventHandle)
+	assert.NoError(t, elog.Close())
+	// Bad event log source name does not cause Open() to fail.
+	// But eventHandle will be 0 and Close() will fail because of it.
+	elog = NewEventLog("FakeBadElogName", LEVELS, GROUP_NAME, STREAM_NAME,
+		RENDER_FMT, DEST, STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.NoError(t, elog.Open())
+	assert.Zero(t, elog.eventHandle)
+	assert.Error(t, elog.Close())
+	// bad LEVELS does not cause Open() to fail.
+	elog = NewEventLog(NAME, []string{"498"}, GROUP_NAME, STREAM_NAME,
+		RENDER_FMT, DEST, STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.NoError(t, elog.Open())
+	assert.NotZero(t, elog.eventHandle)
+	assert.NoError(t, elog.Close())
+	// bad wlog.eventOffset does not cause Open() to fail.
+	elog = NewEventLog(NAME, []string{"498"}, GROUP_NAME, STREAM_NAME,
+		RENDER_FMT, DEST, STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	elog.eventOffset = 9987
+	assert.NoError(t, elog.Open())
+	assert.NotZero(t, elog.eventHandle)
+	assert.NoError(t, elog.Close())
+}
+
+// TestReadGoodSource will verify we can read events written by a registered
+// event log source.
+func TestReadGoodSource(t *testing.T) {
+	elog := NewEventLog(NAME, LEVELS, GROUP_NAME, STREAM_NAME, RENDER_FMT, DEST,
+		STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.NoError(t, elog.Open())
+	seekToEnd(t, elog)
+	writeEvents(t, 10, true, "CWA_UnitTest111", 777)
+	checkEvents(t, elog, []expectedEvent{{"[Application] [ERROR] [777] [CWA_UnitTest111] ", 10}})
+	assert.NoError(t, elog.Close())
+}
+
+// TestReadBadSource will verify that we cannot read events written by an
+// unregistered event log source.
+func TestReadBadSource(t *testing.T) {
+	elog := NewEventLog(NAME, LEVELS, GROUP_NAME, STREAM_NAME, RENDER_FMT, DEST,
+		STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.NoError(t, elog.Open())
+	seekToEnd(t, elog)
+	writeEvents(t, 10, false, "CWA_UnitTest222", 888)
+	checkEvents(t, elog, []expectedEvent{{"CWA_UnitTest222", 0}})
+	assert.NoError(t, elog.Close())
+}
+
+// TestReadWithBothSources will verify we can read events written by a
+// registered event log source, even if the batch contains events from an
+// unregistered source too.
+func TestReadWithBothSources(t *testing.T) {
+	elog := NewEventLog(NAME, LEVELS, GROUP_NAME, STREAM_NAME, RENDER_FMT, DEST,
+		STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.NoError(t, elog.Open())
+	seekToEnd(t, elog)
+	writeEvents(t, 10, true, "CWA_UnitTest111", 777)
+	writeEvents(t, 10, false, "CWA_UnitTest222", 888)
+	ee := []expectedEvent{
+		{"[Application] [ERROR] [777] [CWA_UnitTest111] ", 10},
+		{"CWA_UnitTest222", 0},
+	}
```

Comment:
Ah, I see you had to do it this way because `checkEvents` statefully reads from the event log.  I might have done it like this instead then, for slightly better readability:

```
writeEvents(...)
records := readEvents(t, elog)
checkEvents(t, records, "CWA_UnitTest111", 10)
checkEvents(t, records, "CWA_UnitTest222", 0)
```

But not a big deal.

Reply:
Fixed.